### PR TITLE
Feature/new email validation

### PR DESCRIPTION
**Pull request by craigzour**
# TODO WHEN MERGING

- Replace existing STAGING_NOTIFY_API_KEY (https://github.com/cds-snc/forms-terraform/settings/secrets/actions) with the newly generated one
- Trigger new Terraform deployment
- Create new ticket to keep in mind that the new API Key (Live type) is only a temporary solution so that we can simplify testing but that it should be reverted to the previous one (Team and Safelist type)

# Summary | Résumé

closes https://github.com/cds-snc/platform-forms-client/issues/1375

- Changed government email validation regex
- Added government email validation to form builder settings page

Screen Shot 2022-12-07 at 9 26 13 AM

# Test instructions | Instructions pour tester la modification

- Go to http://localhost:3000/form-builder/settings
- Input an invalid email address. You should see an alert popping up. (See valid domain [here](https://github.com/cds-snc/platform-forms-client/issues/1375))
- Input a valid email address. There should not be any error.

# Pull Request Checklist

Please complete the following items in the checklist before you request a review:

- [x] Have you completely tested the functionality of change introduced in this PR? Is the PR solving the problem it's meant to solve within the scope of the related issue?
- [x] The PR does not introduce any new issues such as failed tests, console warnings or new bugs.
- [x] If this PR adds a package have you ensured its licensed correctly and does not add additional security issues?
- [x] Is the code clean, readable and maintainable? Is it easy to understand and comprehend.
- [x] Does your code have adequate comprehensible comments? Do new functions have [docstrings](https://tsdoc.org/)?
- [x] Have you modified the change log and updated any relevant documentation?
- [x] Is there adequate test coverage? Both unit tests and end-to-end tests where applicable?
- [x] If your PR is touching any UI is it accessible? Have you tested it with a screen reader? Have you tested it with automated testing tools such as axe?

### CODE REVIEW

File: lib/validation.tsx [hunk 1]
```diff
@@ -295,23 +295,12 @@ export const setFocusOnErrorMessage = (props: FormikProps<Responses>, errorId: s
...
-  //Get the domain from email
-  const emailDomain = email.substring(email.lastIndexOf("@") + 1);
-  //Check the email's domain against the list of domains
-  //Search from the end of the string since we care about top-ish level domains most.
-  const domainList = domains.join("|");
-  const regGovDomain = new RegExp(`.?(?:${domainList})$`);
-  return regGovDomain.test(emailDomain);
+export const isValidGovEmail = (email: string): boolean => {
+  const regex =
+    /^([a-zA-Z0-9!#$%&'*+-/=?^_`{|}~.])+@((?:[a-zA-Z0-9-.]+\.gc\.ca|cds-snc\.freshdesk\.com)|(canada|cds-snc|elections|rcafinnovation|canadacouncil)\.ca)$/;
```
[the remark this answers is not in the document]
Comment (craigzour) [reply lib/validation.tsx:302]: I understand but the fact that we have to allow dynamic `*.gc.ca` subdomains as well as having that same validation in both the code and the [JSON validation file](lib/middleware/schemas/templates.schema.json) makes it more consistent to have it implemented this way.

[the remark this answers is not in the document]
Comment (thiessenp-cds) [reply lib/validation.tsx:302]: True that makes sense (insert coder sad tear :)

**Review by timarney — DISMISSED** [review]
@craigzour approving as is but let's create a follow-up ticket to improve a11y 

Currently when the error pops there's no alert / dialog for screen readers

**Comment by timarney** [comment]
Recorded what I see / hear ... it does pop up that text for a split second
https://user-images.githubusercontent.com/62242/206265763-8b41cda2-de55-43d9-bc89-feadfeb9c598.mp4

**Comment by craigzour** [comment]
> Recorded what I see / hear ... it does pop up that text for a split second https://user-images.githubusercontent.com/62242/206265763-8b41cda2-de55-43d9-bc89-feadfeb9c598.mp4

Alright yeah I guess I was lucky when I tested it because it now starts reading the title of that error div and then just stops :(

File: components/form-builder/app/Settings.tsx [hunk 2]
```diff
@@ -69,6 +70,22 @@ const HintText = ({ id, children }: { id: string; children?: JSX.Element | strin
   );
 };
 
+const InvalidEmailError = ({ isActive }: { isActive: boolean }) => {
+  const { t } = useTranslation("form-builder");
+
+  return isActive ? (
+    <div className="mt-2 mb-2" role="alert">
```
Comment (dsamojlenko) [inline components/form-builder/app/Settings.tsx:77]: The alert area should be present on the page before it is filled with content. See this bit in the [aria alert docs](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/alert_role): 

> As with all other [live regions](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions), alerts will only be announced when the content of the element with role="alert" is updated. Make sure that the element with the role is present in the page's markup first

So the outer container with `role="alert"` should be in the dom but invisible, and then populated with the message when triggered.

Comment (dsamojlenko) [reply components/form-builder/app/Settings.tsx:77]: Also could add a `aria-invalid` attribute to the input when the error is triggered https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-invalid

File: components/form-builder/app/Settings.tsx [hunk 3]
```diff
@@ -69,6 +70,22 @@ const HintText = ({ id, children }: { id: string; children?: JSX.Element | strin
...
+      <span className="text-red text-sm font-bold">{t("settingsInvalidEmailAlertTitle")}</span>
```
Comment (dsamojlenko) [inline components/form-builder/app/Settings.tsx:78]: This should probably be a heading. I don't think there are any other headings on the page, so could be H2 and just use tailwind classes to size it appropriately.

**Comment by craigzour** [comment]
PR was updated with recent commits from develop as well as some improvements for accessibility (which are still under testing).

**Comment by timarney** [comment]
🥇 

Screen Shot 2022-12-07 at 3 47 05 PM

**Comment by timarney** [comment]
>  For this reason, it's usually best to briefly "clear" the contents of the alert container before then injecting the alert message.

Wondering if clearing an adding a small delay might help


https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/alert_role

**Review by dsamojlenko — APPROVED** [review]
Nice work, thanks for incorporating the recommended a11y fixes ... we may be able to improve a bit more later.

**Pull request merged**